components/tidb_query/src/codec/convert.rs

```diff
@@ -1692,4 +1828,129 @@ mod tests {
             assert_eq!(res, exp);
         }
     }
+
+    #[test]
+    fn test_produce_str_with_specified_tp() {
+        let cases = vec![
+            // branch 1
+            ("世界，中国", 1, charset::CHARSET_UTF8),
+            ("世界，中国", 2, charset::CHARSET_UTF8),
+            ("世界，中国", 3, charset::CHARSET_UTF8),
+            ("世界，中国", 4, charset::CHARSET_UTF8),
+            ("世界，中国", 5, charset::CHARSET_UTF8),
+            ("世界，中国", 6, charset::CHARSET_UTF8),
+            // branch 2
+            ("世界，中国", 1, charset::CHARSET_ASCII),
+            ("世界，中国", 2, charset::CHARSET_ASCII),
+            ("世界，中国", 3, charset::CHARSET_ASCII),
+            ("世界，中国", 4, charset::CHARSET_ASCII),
+            ("世界，中国", 5, charset::CHARSET_ASCII),
+            ("世界，中国", 6, charset::CHARSET_ASCII),
+        ];
+
+        let mut cfg = EvalConfig::new();
+        cfg.set_flag(Flag::TRUNCATE_AS_WARNING);
+        let mut ctx = EvalContext::new(Arc::new(cfg));
+        let mut ft = FieldType::new();
+
+        for (s, char_num, cs) in cases {
+            ft.set_charset(cs.to_string());
+            ft.set_flen(char_num);
+            let bs = s.as_bytes();
+            let r = produce_str_with_specified_tp(&mut ctx, Cow::Borrowed(bs), &ft, false);
+            assert!(r.is_ok(), "{}, {}, {}", s, char_num, cs);
+            let p = r.unwrap();
+
+            if cs == charset::CHARSET_UTF8MB4 || cs == charset::CHARSET_UTF8 {
+                let ns: String = s.chars().take(char_num as usize).collect();
+                assert_eq!(p.as_ref(), ns.as_bytes(), "{}, {}, {}", s, char_num, cs);
+            } else {
+                assert_eq!(
+                    p.as_ref(),
+                    &bs[..(char_num as usize)],
+                    "{}, {}, {}",
+                    s,
+                    char_num,
+                    cs
+                );
+            }
+        }
+
+        let cases = vec![
+            // branch 3
+            ("世界，中国", 20, charset::CHARSET_ASCII),
+            ("世界，中国", 30, charset::CHARSET_ASCII),
+            ("世界，中国", 50, charset::CHARSET_ASCII),
+        ];
+
+        use tidb_query_datatype::FieldTypeAccessor;
+
+        let mut cfg = EvalConfig::new();
+        cfg.set_flag(Flag::TRUNCATE_AS_WARNING);
+        let mut ctx = EvalContext::new(Arc::new(cfg));
+        let mut ft = FieldType::new();
+
+        <FieldType as FieldTypeAccessor>::set_tp(&mut ft, FieldTypeTp::String);
```

Comment:
`ft.as_accessor().set_tp(...).set_collation(...)`

Reply:
if there is many `.xx()`, then the formator will make one  `.xx()  one line, I think it will a little difficult to read.

Reply:
`FieldTypeAccessor` make chain call easier.